### ISSUE TITLE
indexable arrays<S>  ?

### ISSUE THREAD

**Issue by johnburkey** [realm/realm-cocoa#3669]
Strings are expensive and sometimes projects want to be able to index a hashable array object that has patterns in it and is sortable but isn't actually a string.   It can be stored today as an NSData, but it would be great if it was indexable.

**Comment by mrackwitz** [comment]
I guess, you're speaking here of arrays of strings? We're tracking adding support for storing arrays of primitives (string, int, etc.) as RLMObject properties by issue #1120. Meanwhile you're right, you can use `NSData` as one way to workaround that. Another way would be to create a new model class in your schema. You can then define the string property as indexed.

```swift
class Person : Object {
   dynamic var name = ""
   let petNames = List<PetName>()
}

class PetName : Object {
   dynamic var name = ""

   override static func indexedProperties() -> [String] {
       return ["name"]
   }
}
```

**Comment by johnburkey** [comment]
I would like to use an array of bytes as and indexed property .   It's like a hash except the key is not a string .   Think of it as a piece of music with notes instead .   A pattern from a set but notes instead of letters.    String is too slow for my scenario.   I'm already using Nsdata- but if I could have it be an indexed property it would be more useful .

John

**Comment by mrackwitz** [comment]
If we would allow to index `NSData` properties, they would be indexed as a whole. Indexes can't support arbitrary binary data formats. So if you're searching for a specific note, it wouldn't help you at all. It would only help if you're instead searching for a specific set of notes. So is that what you're looking for?

**Comment by johnburkey** [comment]
Yeah.  Sorry I miscommunicated - that's what I meant.   Just like using a string index as you in effect said .   There is a single value for each byte - as if the byte array were a uint8 backed string .

Thanks, John

**Comment by johnburkey** [comment]
See what's cool about it is it means you can use a byte array backed Nsdata and have indexable enum arrays in effect .   So not arbitrary - the enum gives u homogenous elements and a hash- and you can impose ordering if you like via a protocol.

Thanks, John

**Comment by mrackwitz** [comment]
Ok, I'll promote this to a feature request.

**Comment by johnburkey** [comment]
Thanks a lot!

John

**Comment by johnburkey** [comment]
Did you guys implement this?   I have a super high performant case using byte arrays as a substitute for strings, and want to be able to index them.  Each UInt8 can be handled just like a character- its like an enum.

-  John

**Comment by jpsim** [comment]
No we have not implemented this.

**Comment by johnburkey** [comment]
Alright, well think of it as a feature to do indexing on things that look just like strings, but aren’t actually multi-byte internationally safe unicode savvy things, but are still runs of values that occur in patterns just like characters do in words.   Given how heavy weight string is becoming year after year, its a valuable feature to be able to have variable length piles of symbols / enumerated values in indexable fields in a database.     

Thanks for the reply,

John

**Comment by jpsim** [comment]
Agreed, this is something we want to do, which is why we're tracking it here.

**Comment by johnburkey** [comment]
Sweet.
